In [10]:
# Imports
import numpy as np
import scipy as sp

##Question 1##
Since deterministic noises are defined to be the part of the target function $f$ that the hypothesis set $\mathscr{H}$ can't capture, we can expect that, due to the fact that $\mathscr{H}' \subseteq \mathscr{H}$, more parts of $f$ can't be captured by $\mathscr{H}'$ compared to $\mathscr{H}$. Hence, the deterministic noise will increase.

Therefore, the answer to question 1 is **[b]**.



##Question 2 - 6##

In [11]:
# Load datasets
training_set = np.loadtxt("/content/drive/MyDrive/Colab Notebooks/CS 156a HWs/in.dta.txt", dtype = np.float64)
test_set = np.loadtxt("/content/drive/MyDrive/Colab Notebooks/CS 156a HWs/out.dta.txt")

# Non-linear transform
x_1 = training_set[:,0]
x_2 = training_set[:,1]
trans_mat = np.c_[np.ones(len(x_1)), x_1, x_2, (x_1 ** 2), (x_2 ** 2), x_1 * x_2, abs(x_1 - x_2), abs(x_1 + x_2)]
x_test_1 = test_set[:,0]
x_test_2 = test_set[:,1]
trans_mat_test = np.c_[np.ones(len(x_test_1)), x_test_1, x_test_2, (x_test_1 ** 2), (x_test_2 ** 2), x_test_1 * x_test_2, abs(x_test_1 - x_test_2), abs(x_test_1 + x_test_2)]

# Linear regression
y = training_set[:,2]
y_test = test_set[:,2]

coeff_sol = np.dot(np.matmul(np.linalg.inv(np.matmul(trans_mat.T, trans_mat)), trans_mat.T), y)

# Errors
in_sample = np.mean(np.not_equal(np.sign(np.matmul(trans_mat, coeff_sol)), y))
out_sample = np.mean(np.not_equal(np.sign(np.matmul(trans_mat_test, coeff_sol)), y_test))
print("In-sample error (basic)", in_sample, "Out-of-sample error (basic)", out_sample)

# Apply Weight Decay
k_list = [-3,-2,-1,0,1,2,3]
for k in k_list:
  lam = 10 ** k
  main_body = np.linalg.inv(np.matmul(trans_mat.T, trans_mat) + (lam * np.identity(8)))
  weights = np.matmul(np.matmul(main_body, trans_mat.T), y)
  in_sample = np.mean(np.not_equal(np.sign(np.matmul(trans_mat, weights)), y)) + (lam/len(y)) * np.dot(weights, weights)
  out_sample = np.mean(np.not_equal(np.sign(np.matmul(trans_mat_test, weights)), y_test)) + (lam/len(y_test)) * np.dot(weights, weights)
  print("k =",k)
  print("In-sample error (decayed)", in_sample, "Out-of-sample error (decayed)", out_sample)

In-sample error (basic) 0.02857142857142857 Out-of-sample error (basic) 0.084
k = -3
In-sample error (decayed) 0.029525839138147505 Out-of-sample error (decayed) 0.08013361747934065
k = -2
In-sample error (decayed) 0.03725328025656029 Out-of-sample error (decayed) 0.08521545923591844
k = -1
In-sample error (decayed) 0.07247544093764077 Out-of-sample error (decayed) 0.06214656173126971
k = 0
In-sample error (decayed) 0.07203815960036912 Out-of-sample error (decayed) 0.10208534234405167
k = 1
In-sample error (decayed) 0.20012843656341833 Out-of-sample error (decayed) 0.14401798111887856
k = 2
In-sample error (decayed) 0.2833238906326495 Out-of-sample error (decayed) 0.23966534468857095
k = 3
In-sample error (decayed) 0.38510913211565906 Out-of-sample error (decayed) 0.4379152784961923


Therefore, from the simulation result, the answer to question 2 - 6 are:
2.   **[a]**
3.   **[d]**
4.   **[e]**
5.   **[d]**
6.   **[b]**


## Question 7##
$\mathscr{H}(10, 0, 3)$ is the subset of $\mathscr{H}_{10}$ such that $w_q=0$ for all $q\geq 3$. Hence, only $L_0(x), L_1(x)$ and $L_2(x)$ get non-zero coefficients. Hence, $\mathscr{H}(10, 0, 3) = \{h|h(x) = \sum_{q = 0}^{2}w_qL_q(x) \}= \mathscr{H}_2$.  
For the same reason, $\mathscr{H}(10, 0, 4)$ is the subset of $\mathscr{H}_{10}$ such that only $L_0(x), L_1(x), L_2(x), L_3(x)$ get non-zero coefficients because for any $q\geq 4$, $L_q(x)$ has zero coefficient. Hence, $\mathscr{H}(10, 0, 4) = = \{h|h(x) = \sum_{q = 0}^{3}w_qL_q(x) \}= \mathscr{H}_3$.  
From what we derived above we can see that $\mathscr{H}(10, 0, 3) \subseteq \mathscr{H}(10, 0, 4)$ because we have $\mathscr{H}(10, 0, 3) = \{h(x) = \sum_{q = 0}^{3}w_qL_q(x)\in \mathscr{H}(10, 0, 4)|w_3=0\}$  
Therefore, we get that $\mathscr{H}(10, 0, 3) \cap \mathscr{H}(10, 0, 4) = \mathscr{H}(10, 0, 3) = \mathscr{H}_2$.

Therefore, the answer to this question is **[c]**.

##Question 8 - 10##

For question 8, firstly, for the forward part, from the $0$-th layer to the $1$-st layer, we have $0\leq i \leq 5=d^{(0)}$ and $1 \leq j \leq 3=d^{(1)}$ and from the $1$-st layer to the $2$-nd layer, we have $0\leq i \leq 3=d^{(1)}$ and $1 \leq j \leq 1=d^{(1)}$. Hence, for the operation $w_{ij}^{(l)}x_{i}^{(l-1)}$, the number it is performed is given by:
$$6\times 3 + 4\times 1 = 22$$
Since backward calculations is performed the same number of times as the forward part from layer $2$ to $1$ and we don't need to calculate $\delta$ for the $0$-th layer. Thus, the number of operation $w_{ij}^{(l)}\delta_{j}^{l}$ is
$$4\times 1 = 4$$
And then, since the total number of $w_{ij}^{(l)}$ is equal to the number of forward steps, we know that it's also $22$.
Hence, the total number of operations is
$$22+22+4 = 48 \approx 50$$

Therefore, the answer to question 8 is **[e]**.

For question 9, the minimum can be given by putting each hidden node in a unique layer, then the total number of layer is just as follows:
$$10 + 36 = 46$$

Therefore, the answer to question 9 is **[a]**.

For question 10, since the minimum is obtained by maximizing the number of layers, it's reasonable for us to assume that the maximum can be obtained using a relatively small number of layers. Below is a numerical approach that calculates the maximum number of weights given $1,2$ and $3$ layers:

In [13]:
one_layer_result = 10 * 35 + 36 * 1
two_layer_cases = np.array([[x,y] for x in range(1, 36) for y in range(1, 36) if x + y == 36])
two_layer_best_idx = np.argmax(10 * (two_layer_cases[:,0] - 1) + two_layer_cases[:,0] * (two_layer_cases[:,1] - 1) + two_layer_cases[:,1])
three_layer_cases = np.array([[x,y,z] for x in range(1, 36) for y in range(1, 36) for z in range(1, 36) if x + y + z == 36])
three_layer_best_idx = np.argmax(10 * (three_layer_cases[:,0] - 1) + three_layer_cases[:,0] * (three_layer_cases[:,1] - 1) + three_layer_cases[:,1] * (three_layer_cases[:,2] - 1) + (three_layer_cases[:,2]))

two_layer_best = two_layer_cases[two_layer_best_idx]
best_two = (10 * (two_layer_best[0] - 1)) + (two_layer_best[0] * (two_layer_best[1] - 1)) + two_layer_best[1]
three_layer_best = three_layer_cases[three_layer_best_idx]
best_three = (10 * (three_layer_best[0] - 1)) + (three_layer_best[0] * (three_layer_best[1] - 1)) + (three_layer_best[1] * (three_layer_best[2] - 1)) + three_layer_best[2]

print("Number of weights with one layer:",one_layer_result)
print("Optimal node number and maximum number of weights with two layer:", two_layer_best, best_two)
print("Optimal node number and maximum number of weights with three layer:", three_layer_best, best_three)


Number of weights with one layer: 386
Optimal node number and maximum number of weights with two layer: [22 14] 510
Optimal node number and maximum number of weights with three layer: [22 13  1] 475


From the simulation, it's reasonable to conclude that the maximum is obtained with $2$ layers with the first layer having $22$ nodes and the second layer having $14$ nodes.
  
The resulting number of weights is $510$. Therefore, the answer to question 10 is **[e]**.